### PR TITLE
Prevent redirection loop everytime we rely on HTTP_REFERER

### LINKED ISSUE #256: The page isn't redirecting properly
I'm running an instance of Shaarli 0.0.45-beta on my local desktop computer. If I click on a tag in the main view I get the following error from Iceweasel/Firefox 38.0.1:

> The page isn't redirecting properly. Iceweasel has detected that the server is redirecting the request for this address in a way that will never complete. This problem can sometimes be caused by disabling or refusing to accept cookies.

I also tried with a clean, new Iceweasel profile.

If I move the mouse over the "linux" tag I see the "http://127.0.0.1/shaarli/?addtag=linux" link. But if I click on it I'm redirected to the http://127.0.0.1/shaarli/?addtag=linux&searchtags=linux, which doesn't work.

### PR DESCRIPTION
**Pull request by ArthurHoaro**
* search tag
  * delete tag
  * pagination
  * display privates only
  * delete link

Move location generation to Utils.php + unit tests.

Fixes #256

@fturco: Since I can't reproduce your issue without hard coding values, please test this PR carefully.

### CODE REVIEW

**Comment by nodiscc** [comment]
Thanks. Will merge as soon as I have time to test.

**Comment by ArthurHoaro** [comment]
updated.

**Comment by ArthurHoaro** [comment]
Rebased again:

 * delete forgotten `var_dump`.
 * Fix return URL after adding/editing link with referer spoofing.

**Comment by fturco** [comment]
I still haven't learned Git, I hope the following is correct:

    git clone https://github.com/shaarli/Shaarli.git
    cd Shaarli
    git remote add arthur https://github.com/ArthurHoaro/Shaarli.git
    git fetch arthur
    git checkout tag-http-referer

After installing Shaarli I tried to reproduce the problem I had with redirection but the behaviour was fine. Anyway I have seen two other problems:

* If I edit a bookmark and then click on the Cancel button I am not able to return to main view; it stays on the edit view
* The Filter by tag field is way too long; I need to scroll the page horizontally in order to see the corresponding Search button.

**Comment by ArthurHoaro** [comment]
> If I edit a bookmark and then click on the Cancel button I am not able to return to main view; it stays on the edit view

Yep you're right. I've forgotten `cancel_edit`. You can test the last commit, which should fix that, with `git pull arthur tag-http-referer`.

> The Filter by tag field is way too long; I need to scroll the page horizontally in order to see the corresponding Search button.

That's seems oddly related to #231... which should be fixed. Could you open another issue, since this is not related to HTTP_REFERER and redirections?

**Comment by fturco** [comment]
>> If I edit a bookmark and then click on the Cancel button I am not able to return to main view; it stays on the edit view

> Yep you're right. I've forgotten cancel_edit. You can test the last commit, which should fix that, with git pull arthur tag-http-referer.

Ok, that issue disappeared.

I'm going to check if the other bug has been already reported or otherwise I'm going to report it myself.

File: application/Utils.php [hunk 1]
```diff
@@ -84,4 +84,36 @@ function checkDateFormat($format, $string)
...
+ * Generate a header location from HTTP_REFERER.
+ * Make sure the referrer is Shaarli itself and prevent redirection loop.
+ *
+ * @param string $referrer - HTTP_REFERER.
+ * @param string $host - Server HOST.
+ * @param array $loopTerms - Contains list of term to prevent redirection loop.
+ *
+ * @return string $referer - final referer.
+ */
+function generateLocation($referrer, $host, $loopTerms = array())
```
Comment (virtualtam) [inline application/Utils.php]: typo: `referer` :p

File: index.php [hunk 2]
```diff
@@ -1099,6 +1099,11 @@ function renderPage()
         if (empty($_SERVER['HTTP_REFERER'])) { header('Location: ?searchtags='.urlencode($_GET['addtag'])); exit; } // In case browser does not send HTTP_REFERER
         parse_str(parse_url($_SERVER['HTTP_REFERER'],PHP_URL_QUERY), $params);
 
+        // Prevent redirection loop
+        if(isset($params['addtag'])) {
```
Comment (virtualtam) [inline index.php]: missing space before opening parenthesis

File: index.php [hunk 3]
```diff
@@ -1123,16 +1128,29 @@ function renderPage()
...
+        if (empty($_SERVER['HTTP_REFERER'])) {
+            header('Location: ?');
+            exit;
+        }
+
+        // In case browser does not send HTTP_REFERER
+        parse_str(parse_url($_SERVER['HTTP_REFERER'], PHP_URL_QUERY), $params);
+
+        // Prevent redirection loop
+        if(isset($params['removetag'])) {
```
Comment (virtualtam) [inline index.php]: missing space before opening parenthesis

**Comment by virtualtam** [comment]
works fine, some minor code comments

File: application/Utils.php (hunk 1 above)
Comment (ArthurHoaro) [inline application/Utils.php]: PHPStorm's dictionary don't English good. :)

**Comment by ArthurHoaro** [comment]
Thank you. Fixed and rebased.

**Comment by virtualtam** [comment]
+1
(2 intruder `referrer` left in comments :p)

**Comment by ArthurHoaro** [comment]
Arg! Fixed.

**Comment by virtualtam** [comment]
Thanks!
@fturco, could you please test the latest version and confirm there are no referer-related issue left so we can merge?

**Comment by fturco** [comment]
I can't reproduce the referer-related problem with the latest version. It seems fine to me. You can merge! :)

**Comment by virtualtam** [comment]
Thanks for your feedback! :)

**Pull request merged**